{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d654e17d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello\n"
     ]
    }
   ],
   "source": [
    "! echo 'hello'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "34ea362f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using virtualenv: \u001b[32m/Users/equaglio/Library/Caches/pypoetry/virtualenvs/decider-genetics-QA_7G8QC-py3.13\u001b[39m\n",
      "\u001b[34mUpdating dependencies\u001b[39m\n",
      "\u001b[2K\u001b[34mResolving dependencies...\u001b[39m \u001b[39;2m(2.3s)\u001b[39;22m\n",
      "\n",
      "No dependencies to install or update\n",
      "\n",
      "\u001b[34mWriting lock file\u001b[39m\n",
      "\n",
      "\u001b[39;1mInstalling\u001b[39;22m the current project: \u001b[36mdecider-genetics\u001b[39m (\u001b[39;1m0.1.3\u001b[39;22m)\u001b[1G\u001b[2K\u001b[39;1mInstalling\u001b[39;22m the current project: \u001b[36mdecider-genetics\u001b[39m (\u001b[32m0.1.3\u001b[39m)\n"
     ]
    }
   ],
   "source": [
    "! cd /Users/equaglio/Decider/decider-genetics\n",
    "! rm -f poetry.lock # remove any `poetry.lock` file already present in the folder\n",
    "! poetry env use /Users/equaglio/.pyenv/versions/3.13.2/bin/python3.13 # if current python version is not compatible, you can install a version 3.10+\n",
    "! poetry install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "b558b6b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mUpdating dependencies\u001b[39m\n",
      "\u001b[2K\u001b[34mResolving dependencies...\u001b[39m \u001b[39;2m(0.8s)\u001b[39;22m\n",
      "\n",
      "No dependencies to install or update\n"
     ]
    }
   ],
   "source": [
    "! poetry update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "dde7127a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[39;1mVirtualenv\u001b[39;22m\n",
      "\u001b[34mPython\u001b[39m:         \u001b[32m3.13.2\u001b[39m\n",
      "\u001b[34mImplementation\u001b[39m: \u001b[32mCPython\u001b[39m\n",
      "\u001b[34mPath\u001b[39m:           \u001b[32m/Users/equaglio/Library/Caches/pypoetry/virtualenvs/decider-genetics-QA_7G8QC-py3.13\u001b[39m\n",
      "\u001b[34mExecutable\u001b[39m:     \u001b[32m/Users/equaglio/Library/Caches/pypoetry/virtualenvs/decider-genetics-QA_7G8QC-py3.13/bin/python\u001b[39m\n",
      "\u001b[34mValid\u001b[39m:          \u001b[32mTrue\u001b[39m\n",
      "\n",
      "\u001b[39;1mBase\u001b[39;22m\n",
      "\u001b[34mPlatform\u001b[39m:   \u001b[32mdarwin\u001b[39m\n",
      "\u001b[34mOS\u001b[39m:         \u001b[32mposix\u001b[39m\n",
      "\u001b[34mPython\u001b[39m:     \u001b[32m3.13.2\u001b[39m\n",
      "\u001b[34mPath\u001b[39m:       \u001b[32m/Users/equaglio/.pyenv/versions/3.13.2\u001b[39m\n",
      "\u001b[34mExecutable\u001b[39m: \u001b[32m/Users/equaglio/.pyenv/versions/3.13.2/bin/python3.13\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "! eval $(poetry env activate) # new implementation of `poetry shell`\n",
    "! poetry env info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "0558d1a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/equaglio/Decider/decider-genetics\n",
      "/Users/equaglio/.pyenv/shims/python\n",
      "Dockerfile                \u001b[34mdata\u001b[m\u001b[m                      \u001b[34mold\u001b[m\u001b[m\n",
      "LICENSE                   \u001b[34mdecider_genetics\u001b[m\u001b[m          poetry.lock\n",
      "README.md                 \u001b[34mdocker\u001b[m\u001b[m                    pyproject.toml\n",
      "\u001b[34mbiocypher-log\u001b[m\u001b[m             docker-compose-next.yml   \u001b[34mscripts\u001b[m\u001b[m\n",
      "\u001b[34mbiocypher-out\u001b[m\u001b[m             docker-compose.yml        vignette.ipynb\n",
      "\u001b[34mconfig\u001b[m\u001b[m                    docker-variables.env\n",
      "create_knowledge_graph.py import.report\n",
      "\u001b[33mWarning: [tool.poetry.name] is deprecated. Use [project.name] instead.\u001b[39m\n",
      "\u001b[33mWarning: [tool.poetry.version] is set but 'version' is not in [project.dynamic]. If it is static use [project.version]. If it is dynamic, add 'version' to [project.dynamic].\n",
      "If you want to set the version dynamically via `poetry build --local-version` or you are using a plugin, which sets the version dynamically, you should define the version in [tool.poetry] and add 'version' to [project.dynamic].\u001b[39m\n",
      "\u001b[33mWarning: [tool.poetry.description] is deprecated. Use [project.description] instead.\u001b[39m\n",
      "\u001b[33mWarning: [tool.poetry.readme] is set but 'readme' is not in [project.dynamic]. If it is static use [project.readme]. If it is dynamic, add 'readme' to [project.dynamic].\n",
      "If you want to define multiple readmes, you should define them in [tool.poetry] and add 'readme' to [project.dynamic].\u001b[39m\n",
      "\u001b[33mWarning: [tool.poetry.authors] is deprecated. Use [project.authors] instead.\u001b[39m\n",
      "\u001b[33mWarning: Defining console scripts in [tool.poetry.scripts] is deprecated. Use [project.scripts] instead. ([tool.poetry.scripts] should only be used for scripts of type 'file').\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "! pwd\n",
    "! which python\n",
    "! ls\n",
    "! poetry check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "0f76c0ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36malabaster                    \u001b[39m \u001b[39;1m1.0.0      \u001b[39;22m A light, configurable Sphinx theme\n",
      "\u001b[36mannotated-types              \u001b[39m \u001b[39;1m0.7.0      \u001b[39;22m Reusable constraint types to use ...\n",
      "\u001b[36mappdirs                      \u001b[39m \u001b[39;1m1.4.4      \u001b[39;22m A small Python module for determi...\n",
      "\u001b[36mappnope                      \u001b[39m \u001b[39;1m0.1.4      \u001b[39;22m Disable App Nap on macOS >= 10.9\n",
      "\u001b[36masttokens                    \u001b[39m \u001b[39;1m3.0.0      \u001b[39;22m Annotate AST trees with source co...\n",
      "\u001b[36mattrs                        \u001b[39m \u001b[39;1m25.3.0     \u001b[39;22m Classes Without Boilerplate\n",
      "\u001b[36mbabel                        \u001b[39m \u001b[39;1m2.17.0     \u001b[39;22m Internationalization utilities\n",
      "\u001b[36mbiocypher                    \u001b[39m \u001b[39;1m0.9.3      \u001b[39;22m A unifying framework for biomedic...\n",
      "\u001b[36mblack                        \u001b[39m \u001b[39;1m25.1.0     \u001b[39;22m The uncompromising code formatter.\n",
      "\u001b[36mcertifi                      \u001b[39m \u001b[39;1m2025.4.26  \u001b[39;22m Python package for providing Mozi...\n",
      "\u001b[36mchardet                      \u001b[39m \u001b[39;1m5.2.0      \u001b[39;22m Universal encoding detector for P...\n",
      "\u001b[36mcharset-normalizer           \u001b[39m \u001b[39;1m3.4.1      \u001b[39;22m The Real First Universal Charset ...\n",
      "\u001b[36mclick                        \u001b[39m \u001b[39;1m8.1.8      \u001b[39;22m Composable command line interface...\n",
      "\u001b[36mcolorlog                     \u001b[39m \u001b[39;1m6.9.0      \u001b[39;22m Add colours to the output of Pyth...\n",
      "\u001b[36mcomm                         \u001b[39m \u001b[39;1m0.2.2      \u001b[39;22m Jupyter Python Comm implementatio...\n",
      "\u001b[36mdebugpy                      \u001b[39m \u001b[39;1m1.8.14     \u001b[39;22m An implementation of the Debug Ad...\n",
      "\u001b[36mdecorator                    \u001b[39m \u001b[39;1m5.2.1      \u001b[39;22m Decorators for Humans\n",
      "\u001b[36mdocutils                     \u001b[39m \u001b[39;1m0.21.2     \u001b[39;22m Docutils -- Python Documentation ...\n",
      "\u001b[36mexecuting                    \u001b[39m \u001b[39;1m2.2.0      \u001b[39;22m Get the currently executing AST n...\n",
      "\u001b[36mfrictionless                 \u001b[39m \u001b[39;1m4.40.8     \u001b[39;22m Data management framework for Pyt...\n",
      "\u001b[36midna                         \u001b[39m \u001b[39;1m3.10       \u001b[39;22m Internationalized Domain Names in...\n",
      "\u001b[36mimagesize                    \u001b[39m \u001b[39;1m1.4.1      \u001b[39;22m Getting image size from png/jpeg/...\n",
      "\u001b[36miniconfig                    \u001b[39m \u001b[39;1m2.1.0      \u001b[39;22m brain-dead simple config-ini parsing\n",
      "\u001b[36mipykernel                    \u001b[39m \u001b[39;1m6.29.5     \u001b[39;22m IPython Kernel for Jupyter\n",
      "\u001b[36mipython                      \u001b[39m \u001b[39;1m9.2.0      \u001b[39;22m IPython: Productive Interactive C...\n",
      "\u001b[36mipython-pygments-lexers      \u001b[39m \u001b[39;1m1.1.1      \u001b[39;22m Defines a variety of Pygments lex...\n",
      "\u001b[36misodate                      \u001b[39m \u001b[39;1m0.6.1      \u001b[39;22m An ISO 8601 date/time/duration pa...\n",
      "\u001b[36mjedi                         \u001b[39m \u001b[39;1m0.19.2     \u001b[39;22m An autocompletion tool for Python...\n",
      "\u001b[36mjinja2                       \u001b[39m \u001b[39;1m3.1.6      \u001b[39;22m A very fast and expressive templa...\n",
      "\u001b[36mjsonargparse                 \u001b[39m \u001b[39;1m4.39.0     \u001b[39;22m Implement minimal boilerplate CLI...\n",
      "\u001b[36mjsonschema                   \u001b[39m \u001b[39;1m4.23.0     \u001b[39;22m An implementation of JSON Schema ...\n",
      "\u001b[36mjsonschema-specifications    \u001b[39m \u001b[39;1m2025.4.1   \u001b[39;22m The JSON Schema meta-schemas and ...\n",
      "\u001b[36mjupyter-client               \u001b[39m \u001b[39;1m8.6.3      \u001b[39;22m Jupyter protocol implementation a...\n",
      "\u001b[36mjupyter-core                 \u001b[39m \u001b[39;1m5.7.2      \u001b[39;22m Jupyter core package. A base pack...\n",
      "\u001b[36mmarkdown-it-py               \u001b[39m \u001b[39;1m3.0.0      \u001b[39;22m Python port of markdown-it. Markd...\n",
      "\u001b[36mmarko                        \u001b[39m \u001b[39;1m2.1.3      \u001b[39;22m A markdown parser with high exten...\n",
      "\u001b[36mmarkupsafe                   \u001b[39m \u001b[39;1m3.0.2      \u001b[39;22m Safely add untrusted strings to H...\n",
      "\u001b[36mmatplotlib-inline            \u001b[39m \u001b[39;1m0.1.7      \u001b[39;22m Inline Matplotlib backend for Jup...\n",
      "\u001b[36mmdurl                        \u001b[39m \u001b[39;1m0.1.2      \u001b[39;22m Markdown URL utilities\n",
      "\u001b[36mmore-itertools               \u001b[39m \u001b[39;1m10.7.0     \u001b[39;22m More routines for operating on it...\n",
      "\u001b[36mmypy-extensions              \u001b[39m \u001b[39;1m1.1.0      \u001b[39;22m Type system extensions for progra...\n",
      "\u001b[36mneo4j                        \u001b[39m \u001b[39;1m4.4.12     \u001b[39;22m Neo4j Bolt driver for Python\n",
      "\u001b[36mneo4j-utils                  \u001b[39m \u001b[39;1m0.0.7      \u001b[39;22m \n",
      "\u001b[36mnest-asyncio                 \u001b[39m \u001b[39;1m1.6.0      \u001b[39;22m Patch asyncio to allow nested eve...\n",
      "\u001b[36mnetworkx                     \u001b[39m \u001b[39;1m3.4.2      \u001b[39;22m Python package for creating and m...\n",
      "\u001b[36mnumpy                        \u001b[39m \u001b[39;1m2.2.5      \u001b[39;22m Fundamental package for array com...\n",
      "\u001b[36mowlready2                    \u001b[39m \u001b[39;1m0.47       \u001b[39;22m A package for ontology-oriented p...\n",
      "\u001b[36mpackaging                    \u001b[39m \u001b[39;1m25.0       \u001b[39;22m Core utilities for Python packages\n",
      "\u001b[36mpandas                       \u001b[39m \u001b[39;1m2.2.3      \u001b[39;22m Powerful data structures for data...\n",
      "\u001b[36mpandera                      \u001b[39m \u001b[39;1m0.23.1     \u001b[39;22m A light-weight and flexible data ...\n",
      "\u001b[36mparso                        \u001b[39m \u001b[39;1m0.8.4      \u001b[39;22m A Python Parser\n",
      "\u001b[36mpathspec                     \u001b[39m \u001b[39;1m0.12.1     \u001b[39;22m Utility library for gitignore sty...\n",
      "\u001b[36mpetl                         \u001b[39m \u001b[39;1m1.7.16     \u001b[39;22m A Python package for extracting, ...\n",
      "\u001b[36mpexpect                      \u001b[39m \u001b[39;1m4.9.0      \u001b[39;22m Pexpect allows easy control of in...\n",
      "\u001b[36mplatformdirs                 \u001b[39m \u001b[39;1m4.3.7      \u001b[39;22m A small Python package for determ...\n",
      "\u001b[36mpluggy                       \u001b[39m \u001b[39;1m1.5.0      \u001b[39;22m plugin and hook calling mechanism...\n",
      "\u001b[36mpooch                        \u001b[39m \u001b[39;1m1.8.2      \u001b[39;22m A friend to fetch your data files\n",
      "\u001b[36mprompt-toolkit               \u001b[39m \u001b[39;1m3.0.51     \u001b[39;22m Library for building powerful int...\n",
      "\u001b[36mpsutil                       \u001b[39m \u001b[39;1m7.0.0      \u001b[39;22m Cross-platform lib for process an...\n",
      "\u001b[36mptyprocess                   \u001b[39m \u001b[39;1m0.7.0      \u001b[39;22m Run a subprocess in a pseudo term...\n",
      "\u001b[36mpure-eval                    \u001b[39m \u001b[39;1m0.2.3      \u001b[39;22m Safely evaluate AST nodes without...\n",
      "\u001b[36mpydantic                     \u001b[39m \u001b[39;1m2.11.3     \u001b[39;22m Data validation using Python type...\n",
      "\u001b[36mpydantic-core                \u001b[39m \u001b[39;1m2.33.1     \u001b[39;22m Core functionality for Pydantic v...\n",
      "\u001b[36mpygments                     \u001b[39m \u001b[39;1m2.19.1     \u001b[39;22m Pygments is a syntax highlighting...\n",
      "\u001b[36mpyparsing                    \u001b[39m \u001b[39;1m3.2.3      \u001b[39;22m pyparsing module - Classes and me...\n",
      "\u001b[36mpytest                       \u001b[39m \u001b[39;1m8.3.5      \u001b[39;22m pytest: simple powerful testing w...\n",
      "\u001b[36mpython-dateutil              \u001b[39m \u001b[39;1m2.9.0.post0\u001b[39;22m Extensions to the standard Python...\n",
      "\u001b[36mpython-slugify               \u001b[39m \u001b[39;1m8.0.4      \u001b[39;22m A Python slugify application that...\n",
      "\u001b[36mpytz                         \u001b[39m \u001b[39;1m2025.2     \u001b[39;22m World timezone definitions, moder...\n",
      "\u001b[36mpyyaml                       \u001b[39m \u001b[39;1m6.0.2      \u001b[39;22m YAML parser and emitter for Python\n",
      "\u001b[36mpyzmq                        \u001b[39m \u001b[39;1m26.4.0     \u001b[39;22m Python bindings for 0MQ\n",
      "\u001b[36mrdflib                       \u001b[39m \u001b[39;1m6.3.2      \u001b[39;22m RDFLib is a Python library for wo...\n",
      "\u001b[36mreferencing                  \u001b[39m \u001b[39;1m0.36.2     \u001b[39;22m JSON Referencing + Python\n",
      "\u001b[36mrequests                     \u001b[39m \u001b[39;1m2.32.3     \u001b[39;22m Python HTTP for Humans.\n",
      "\u001b[36mrfc3986                      \u001b[39m \u001b[39;1m2.0.0      \u001b[39;22m Validating URI References per RFC...\n",
      "\u001b[36mrich                         \u001b[39m \u001b[39;1m14.0.0     \u001b[39;22m Render rich text, tables, progres...\n",
      "\u001b[36mroman-numerals-py            \u001b[39m \u001b[39;1m3.1.0      \u001b[39;22m Manipulate well-formed Roman nume...\n",
      "\u001b[36mrpds-py                      \u001b[39m \u001b[39;1m0.24.0     \u001b[39;22m Python bindings to Rust's persist...\n",
      "\u001b[36mshellingham                  \u001b[39m \u001b[39;1m1.5.4      \u001b[39;22m Tool to Detect Surrounding Shell\n",
      "\u001b[36msimpleeval                   \u001b[39m \u001b[39;1m1.0.3      \u001b[39;22m A simple, safe single expression ...\n",
      "\u001b[36msix                          \u001b[39m \u001b[39;1m1.17.0     \u001b[39;22m Python 2 and 3 compatibility util...\n",
      "\u001b[36msnowballstemmer              \u001b[39m \u001b[39;1m2.2.0      \u001b[39;22m This package provides 29 stemmers...\n",
      "\u001b[36msphinx                       \u001b[39m \u001b[39;1m8.2.3      \u001b[39;22m Python documentation generator\n",
      "\u001b[36msphinx-rtd-theme             \u001b[39m \u001b[39;1m3.0.2      \u001b[39;22m Read the Docs theme for Sphinx\n",
      "\u001b[36msphinxcontrib-applehelp      \u001b[39m \u001b[39;1m2.0.0      \u001b[39;22m sphinxcontrib-applehelp is a Sphi...\n",
      "\u001b[36msphinxcontrib-devhelp        \u001b[39m \u001b[39;1m2.0.0      \u001b[39;22m sphinxcontrib-devhelp is a sphinx...\n",
      "\u001b[36msphinxcontrib-htmlhelp       \u001b[39m \u001b[39;1m2.1.0      \u001b[39;22m sphinxcontrib-htmlhelp is a sphin...\n",
      "\u001b[36msphinxcontrib-jquery         \u001b[39m \u001b[39;1m4.1        \u001b[39;22m Extension to include jQuery on ne...\n",
      "\u001b[36msphinxcontrib-jsmath         \u001b[39m \u001b[39;1m1.0.1      \u001b[39;22m A sphinx extension which renders ...\n",
      "\u001b[36msphinxcontrib-qthelp         \u001b[39m \u001b[39;1m2.0.0      \u001b[39;22m sphinxcontrib-qthelp is a sphinx ...\n",
      "\u001b[36msphinxcontrib-serializinghtml\u001b[39m \u001b[39;1m2.0.0      \u001b[39;22m sphinxcontrib-serializinghtml is ...\n",
      "\u001b[36mstack-data                   \u001b[39m \u001b[39;1m0.6.3      \u001b[39;22m Extract data from python stack fr...\n",
      "\u001b[36mstringcase                   \u001b[39m \u001b[39;1m1.2.0      \u001b[39;22m String case converter.\n",
      "\u001b[36mtabulate                     \u001b[39m \u001b[39;1m0.9.0      \u001b[39;22m Pretty-print tabular data\n",
      "\u001b[36mtext-unidecode               \u001b[39m \u001b[39;1m1.3        \u001b[39;22m The most basic Text::Unidecode port\n",
      "\u001b[36mtoml                         \u001b[39m \u001b[39;1m0.10.2     \u001b[39;22m Python Library for Tom's Obvious,...\n",
      "\u001b[36mtornado                      \u001b[39m \u001b[39;1m6.4.2      \u001b[39;22m Tornado is a Python web framework...\n",
      "\u001b[36mtqdm                         \u001b[39m \u001b[39;1m4.67.1     \u001b[39;22m Fast, Extensible Progress Meter\n",
      "\u001b[36mtraitlets                    \u001b[39m \u001b[39;1m5.14.3     \u001b[39;22m Traitlets Python configuration sy...\n",
      "\u001b[36mtreelib                      \u001b[39m \u001b[39;1m1.6.4      \u001b[39;22m A Python 2/3 implementation of tr...\n",
      "\u001b[36mtypeguard                    \u001b[39m \u001b[39;1m4.4.2      \u001b[39;22m Run-time type checker for Python\n",
      "\u001b[36mtyper                        \u001b[39m \u001b[39;1m0.15.3     \u001b[39;22m Typer, build great CLIs. Easy to ...\n",
      "\u001b[36mtyping-extensions            \u001b[39m \u001b[39;1m4.13.2     \u001b[39;22m Backported and Experimental Type ...\n",
      "\u001b[36mtyping-inspect               \u001b[39m \u001b[39;1m0.9.0      \u001b[39;22m Runtime inspection utilities for ...\n",
      "\u001b[36mtyping-inspection            \u001b[39m \u001b[39;1m0.4.0      \u001b[39;22m Runtime typing introspection tools\n",
      "\u001b[36mtzdata                       \u001b[39m \u001b[39;1m2025.2     \u001b[39;22m Provider of IANA time zone data\n",
      "\u001b[36murllib3                      \u001b[39m \u001b[39;1m2.4.0      \u001b[39;22m HTTP library with thread-safe con...\n",
      "\u001b[36mvalidators                   \u001b[39m \u001b[39;1m0.34.0     \u001b[39;22m Python Data Validation for Humans™\n",
      "\u001b[36mwcwidth                      \u001b[39m \u001b[39;1m0.2.13     \u001b[39;22m Measures the displayed width of u...\n",
      "\u001b[36mxdg-base-dirs                \u001b[39m \u001b[39;1m6.0.2      \u001b[39;22m Variables defined by the XDG Base...\n"
     ]
    }
   ],
   "source": [
    "! poetry show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "f64f6094",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "my_mapping_1_variants.yaml my_mapping_3_clinical.yaml\n",
      "my_mapping_2_clinical.yaml my_mapping_3_cna.yaml\n",
      "my_mapping_2_variants.yaml my_mapping_3_variants.yaml\n"
     ]
    }
   ],
   "source": [
    "# mapping files\n",
    "! ls -R decider_genetics/adapters/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "457d7502-9c77-4dfa-b3fc-b51fa928932b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/equaglio/Library/Caches/pypoetry/virtualenvs/decider-genetics-QA_7G8QC-py3.13/bin/python\n",
      "/Users/equaglio/.pyenv/shims/python\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "print(sys.executable)\n",
    "! which python\n",
    "#sys.path.append(\"/Users/equaglio/Decider/Ontoweaver/src\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "57fa595c-bc29-459a-9ffa-b3c4646e724b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following packages are already present in the pyproject.toml and will be skipped:\n",
      "\n",
      "  - \u001b[36mipykernel\u001b[39m\n",
      "\n",
      "If you want to update it to the latest compatible version, you can use `poetry update package`.\n",
      "If you prefer to upgrade it to the latest available version, you can use `poetry add package@latest`.\n",
      "\n",
      "Nothing to add.\n"
     ]
    }
   ],
   "source": [
    "! poetry add ipykernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "e4f06018-4478-4842-a7d7-a1fb857d9b84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Installed kernelspec my-poetry-env in /Users/equaglio/Library/Jupyter/kernels/my-poetry-env\n"
     ]
    }
   ],
   "source": [
    "# to point the kernel of the notebook to the poetry environment activated. \n",
    "! poetry run python -m ipykernel install --user --name=my-poetry-env --display-name \"decider genetics poetry env\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b7903f2-9407-4766-b246-645b83c53e56",
   "metadata": {},
   "source": [
    "### Now change kernel and make sure to use \"decider genetics poetry env\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e818ac1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO -- This is BioCypher v0.9.3.\n",
      "INFO -- Logging into `biocypher-log/biocypher-20250429-183337.log`.\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "import logging\n",
    "import pandas as pd\n",
    "import yaml\n",
    "from biocypher import BioCypher\n",
    "import ontoweaver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "87b71308",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34moncodash files\u001b[m\u001b[m                         synthetic_cna.csv\n",
      "oncokb_biomarker_drug_associations.tsv synthetic_variants.csv\n",
      "synthetic_clinical.csv\n",
      "\n",
      "data/oncodash files:\n",
      "BiologicalProcess-part000.csv       GeneToBiologicalProcess-part000.csv\n"
     ]
    }
   ],
   "source": [
    "# tables in data/\n",
    "! ls -R data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0e6a0503",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ID</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Func.MANE</th>\n",
       "      <th>Gene.MANE</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomAD_genome_FIN</th>\n",
       "      <th>gnomAD_genome_max</th>\n",
       "      <th>gnomAD_exome_nc_ALL</th>\n",
       "      <th>gnomAD_exome_nc_NFE</th>\n",
       "      <th>gnomAD_exome_nc_NFE_SWE</th>\n",
       "      <th>gnomAD_exome_nc_FIN</th>\n",
       "      <th>gnomAD_exome_nc_max</th>\n",
       "      <th>Truncal</th>\n",
       "      <th>readCounts</th>\n",
       "      <th>samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>patient1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>32660776</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>8p12</td>\n",
       "      <td>intronic</td>\n",
       "      <td>NRG1</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>heterogeneous</td>\n",
       "      <td>38,0;21,0;26,13;37,1</td>\n",
       "      <td>patient1_Per</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>patient1</td>\n",
       "      <td>chr20</td>\n",
       "      <td>40688333</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>20q12</td>\n",
       "      <td>exonic</td>\n",
       "      <td>MAFB</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>homogeneous</td>\n",
       "      <td>65,0;70,17;30,5;62,9;58,7</td>\n",
       "      <td>patient1_Lum;patient1_Per</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>patient1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>197391056</td>\n",
       "      <td>CAGTAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>2q33.1</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>SF3B1</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>heterogeneous</td>\n",
       "      <td>30,0;29,17;77,0;60,0;44,0</td>\n",
       "      <td>patient1_Per;patient1_Ome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>patient1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>124089869</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Xq25</td>\n",
       "      <td>intronic</td>\n",
       "      <td>STAG2</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>heterogeneous</td>\n",
       "      <td>34,0;33,0;45,0;22,3;28,7</td>\n",
       "      <td>patient1_Asc;patient1_Lum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>patient1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>151916206</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs977154062</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6q25.1</td>\n",
       "      <td>intronic</td>\n",
       "      <td>ESR1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>9.52064e-05</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>homogeneous</td>\n",
       "      <td>58,0;21,6</td>\n",
       "      <td>patient1_Lum;patient1_Asc;patient1_Per;patient...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    patient  CHROM        POS      REF ALT           ID FILTER cytoBand  \\\n",
       "0  patient1   chr8   32660776        T   A            .   PASS     8p12   \n",
       "1  patient1  chr20   40688333        G   C            .   PASS    20q12   \n",
       "2  patient1   chr2  197391056  CAGTAAA   C            .   PASS   2q33.1   \n",
       "3  patient1   chrX  124089869        G   T            .   PASS     Xq25   \n",
       "4  patient1   chr6  151916206        G   A  rs977154062   PASS   6q25.1   \n",
       "\n",
       "  Func.MANE Gene.MANE  ... gnomAD_genome_FIN gnomAD_genome_max  \\\n",
       "0  intronic      NRG1  ...                 .                 .   \n",
       "1    exonic      MAFB  ...                 .                 .   \n",
       "2      UTR3     SF3B1  ...                 .                 .   \n",
       "3  intronic     STAG2  ...                 .                 .   \n",
       "4  intronic      ESR1  ...                 0       9.52064e-05   \n",
       "\n",
       "  gnomAD_exome_nc_ALL gnomAD_exome_nc_NFE gnomAD_exome_nc_NFE_SWE  \\\n",
       "0                   .                   .                       .   \n",
       "1                   .                   .                       .   \n",
       "2                   .                   .                       .   \n",
       "3                   .                   .                       .   \n",
       "4                   .                   .                       .   \n",
       "\n",
       "  gnomAD_exome_nc_FIN gnomAD_exome_nc_max        Truncal  \\\n",
       "0                   .                   .  heterogeneous   \n",
       "1                   .                   .    homogeneous   \n",
       "2                   .                   .  heterogeneous   \n",
       "3                   .                   .  heterogeneous   \n",
       "4                   .                   .    homogeneous   \n",
       "\n",
       "                  readCounts  \\\n",
       "0       38,0;21,0;26,13;37,1   \n",
       "1  65,0;70,17;30,5;62,9;58,7   \n",
       "2  30,0;29,17;77,0;60,0;44,0   \n",
       "3   34,0;33,0;45,0;22,3;28,7   \n",
       "4                  58,0;21,6   \n",
       "\n",
       "                                             samples  \n",
       "0                                       patient1_Per  \n",
       "1                          patient1_Lum;patient1_Per  \n",
       "2                          patient1_Per;patient1_Ome  \n",
       "3                          patient1_Asc;patient1_Lum  \n",
       "4  patient1_Lum;patient1_Asc;patient1_Per;patient...  \n",
       "\n",
       "[5 rows x 49 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_df=pd.read_table('data/synthetic_variants.csv')\n",
    "variants_df.head(n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c2c4d74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['patient', 'CHROM', 'POS', 'REF', 'ALT', 'ID', 'FILTER', 'cytoBand',\n",
       "       'Func.MANE', 'Gene.MANE', 'GeneDetail.MANE', 'ExonicFunc.MANE',\n",
       "       'AAChange.MANE', 'Func.refGene', 'Gene.refGene', 'GeneDetail.refGene',\n",
       "       'ExonicFunc.refGene', 'AAChange.refGene', 'genomicSuperDups',\n",
       "       'dbscSNV_ADA_SCORE', 'dbscSNV_RF_SCORE', 'COSMIC_ID',\n",
       "       'COSMIC_OCCURRENCE', 'COSMIC_TOTAL_OCC', 'COSMIC_CONF_SOMA', 'CLNSIG',\n",
       "       'CLNSIGCONF', 'CLNDN', 'CLNREVSTAT', 'CLNALLELEID', 'CLNDISDB',\n",
       "       'Interpro_domain', 'regulomeDB', 'CADD_raw', 'CADD_phred', '1000G_ALL',\n",
       "       '1000G_EUR', 'gnomAD_genome_ALL', 'gnomAD_genome_NFE',\n",
       "       'gnomAD_genome_FIN', 'gnomAD_genome_max', 'gnomAD_exome_nc_ALL',\n",
       "       'gnomAD_exome_nc_NFE', 'gnomAD_exome_nc_NFE_SWE', 'gnomAD_exome_nc_FIN',\n",
       "       'gnomAD_exome_nc_max', 'Truncal', 'readCounts', 'samples'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce19cb74-d8d9-4df6-a7e9-3b539feaa1d5",
   "metadata": {},
   "source": [
    "### Here we show how to write a proper schema config"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7f7f131-b17a-4b9c-b67d-fa2dec27f61e",
   "metadata": {},
   "source": [
    "### Here we show how to write a proper mapping\n",
    "TODO: highlight and show problem of consistency of direction between predicate as defined in the ontology (pointed to in the biocypher_config file) and subject-object assignment in the ontoweaver mapping:\n",
    "    \n",
    "    Issues:\n",
    "* apparently biocypher has no way to declare inverse property\n",
    "* it's not clear to me if and why (conceptually) we shouldn't be able to treat rowIndex as object and another column has suject (other than forcing the user to always have a subject when they define the mapping).\n",
    "* we could add an inverse property on our ontology for example with protege as we create subclasses through biocypher, but personalizing the ontology through 2 different tools is not as convenient/easy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1a4749b8-7867-4140-8cc8-0b5527156353",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edit the biocypher config to point to the new schema_config file\n",
    "\n",
    "with open('config/biocypher_config.yaml', 'r') as f:\n",
    "    bc = yaml.safe_load(f)\n",
    "\n",
    "bc['biocypher']['schema_config_path'] = 'schema_config_1.yaml'\n",
    "\n",
    "with open('config/biocypher_config.yaml', 'w') as f:\n",
    "    yaml.dump(bc, f, default_flow_style=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f22530f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO -- Running BioCypher with schema configuration from config/schema_config_1.yaml.\n",
      "INFO:biocypher:Running BioCypher with schema configuration from config/schema_config_1.yaml.\n",
      "INFO -- Running BioCypher with schema configuration from config/schema_config_1.yaml.\n",
      "INFO:biocypher:Running BioCypher with schema configuration from config/schema_config_1.yaml.\n",
      "INFO -- Loading ontologies...\n",
      "INFO:biocypher:Loading ontologies...\n",
      "INFO -- Instantiating OntologyAdapter class for https://github.com/biolink/biolink-model/raw/v3.2.1/biolink-model.owl.ttl.\n",
      "INFO:biocypher:Instantiating OntologyAdapter class for https://github.com/biolink/biolink-model/raw/v3.2.1/biolink-model.owl.ttl.\n",
      "INFO -- Creating output directory `/Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129`.\n",
      "INFO:biocypher:Creating output directory `/Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129`.\n",
      "DEBUG:biocypher:Writing node CSV from generator.\n",
      "DEBUG:biocypher:Translating nodes to BioCypher\n",
      "ERROR -- No ontology type defined for `Gene`: ABL1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ABL1\n",
      "ERROR -- No ontology type defined for `Gene`: JAK2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: JAK2\n",
      "ERROR -- No ontology type defined for `Gene`: ZRSR2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ZRSR2\n",
      "ERROR -- No ontology type defined for `Gene`: CEBPA\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: CEBPA\n",
      "ERROR -- No ontology type defined for `Gene`: BARD1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: BARD1\n",
      "ERROR -- No ontology type defined for `Gene`: IL7R\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: IL7R\n",
      "ERROR -- No ontology type defined for `Gene`: PHF6\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: PHF6\n",
      "ERROR -- No ontology type defined for `Gene`: KLF2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: KLF2\n",
      "ERROR -- No ontology type defined for `Gene`: MDM2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: MDM2\n",
      "ERROR -- No ontology type defined for `Gene`: NRG1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: NRG1\n",
      "ERROR -- No ontology type defined for `Gene`: MAF\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: MAF\n",
      "ERROR -- No ontology type defined for `Gene`: ASXL1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ASXL1\n",
      "ERROR -- No ontology type defined for `Gene`: CHEK2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: CHEK2\n",
      "ERROR -- No ontology type defined for `Gene`: NOTCH1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: NOTCH1\n",
      "ERROR -- No ontology type defined for `Gene`: BRCA1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: BRCA1\n",
      "ERROR -- No ontology type defined for `Gene`: NTRK3\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: NTRK3\n",
      "ERROR -- No ontology type defined for `Gene`: TCF3\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: TCF3\n",
      "ERROR -- No ontology type defined for `Gene`: STAG2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: STAG2\n",
      "ERROR -- No ontology type defined for `Gene`: BCL6\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: BCL6\n",
      "ERROR -- No ontology type defined for `Gene`: ERBB2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ERBB2\n",
      "ERROR -- No ontology type defined for `Gene`: SETBP1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: SETBP1\n",
      "ERROR -- No ontology type defined for `Gene`: EP300\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: EP300\n",
      "ERROR -- No ontology type defined for `Gene`: CRLF2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: CRLF2\n",
      "ERROR -- No ontology type defined for `Gene`: DUSP22\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: DUSP22\n",
      "ERROR -- No ontology type defined for `Gene`: BRAF\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: BRAF\n",
      "ERROR -- No ontology type defined for `Gene`: ESR1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ESR1\n",
      "ERROR -- No ontology type defined for `Gene`: CDKN2A\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: CDKN2A\n",
      "ERROR -- No ontology type defined for `Gene`: KRAS\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: KRAS\n",
      "ERROR -- No ontology type defined for `Gene`: ATM\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ATM\n",
      "ERROR -- No ontology type defined for `Gene`: MET\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: MET\n",
      "ERROR -- No ontology type defined for `Gene`: PDCD1LG2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: PDCD1LG2\n",
      "ERROR -- No ontology type defined for `Gene`: CD274\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: CD274\n",
      "ERROR -- No ontology type defined for `Gene`: BCOR\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: BCOR\n",
      "ERROR -- No ontology type defined for `Gene`: IDH2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: IDH2\n",
      "ERROR -- No ontology type defined for `Gene`: HRAS\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: HRAS\n",
      "ERROR -- No ontology type defined for `Gene`: BIRC3\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: BIRC3\n",
      "ERROR -- No ontology type defined for `Gene`: TYK2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: TYK2\n",
      "ERROR -- No ontology type defined for `Gene`: FANCA\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: FANCA\n",
      "ERROR -- No ontology type defined for `Gene`: MAFB\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: MAFB\n",
      "ERROR -- No ontology type defined for `Gene`: ALK\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ALK\n",
      "ERROR -- No ontology type defined for `Gene`: FGFR3\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: FGFR3\n",
      "ERROR -- No ontology type defined for `Gene`: ATRX\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ATRX\n",
      "ERROR -- No ontology type defined for `Gene`: DNMT3A\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: DNMT3A\n",
      "ERROR -- No ontology type defined for `Gene`: PDGFB\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: PDGFB\n",
      "ERROR -- No ontology type defined for `Gene`: IDH1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: IDH1\n",
      "ERROR -- No ontology type defined for `Gene`: SF3B1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: SF3B1\n",
      "ERROR -- No ontology type defined for `Gene`: PIK3CA\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: PIK3CA\n",
      "ERROR -- No ontology type defined for `Gene`: TP63\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: TP63\n",
      "ERROR -- No ontology type defined for `Gene`: CALR\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: CALR\n",
      "ERROR -- No ontology type defined for `Gene`: FBXW7\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: FBXW7\n",
      "ERROR -- No ontology type defined for `Gene`: MAP2K1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: MAP2K1\n",
      "ERROR -- No ontology type defined for `Gene`: FGFR2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: FGFR2\n",
      "ERROR -- No ontology type defined for `Gene`: FGFR1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: FGFR1\n",
      "ERROR -- No ontology type defined for `Gene`: IRF4\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: IRF4\n",
      "ERROR -- No ontology type defined for `Gene`: ERCC2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ERCC2\n",
      "ERROR -- No ontology type defined for `Gene`: ETNK1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: ETNK1\n",
      "ERROR -- No ontology type defined for `Gene`: EZH2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: EZH2\n",
      "ERROR -- No ontology type defined for `Gene`: SMARCB1\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: SMARCB1\n",
      "ERROR -- No ontology type defined for `Gene`: BCL2\n",
      "ERROR:biocypher:No ontology type defined for `Gene`: BCL2\n",
      "DEBUG:biocypher:Finished translating nodes to BioCypher.\n",
      "INFO -- Writing 650 entries to SequenceVariant-part000.csv\n",
      "INFO:biocypher:Writing 650 entries to SequenceVariant-part000.csv\n",
      "INFO -- Writing 20 entries to Patient-part000.csv\n",
      "INFO:biocypher:Writing 20 entries to Patient-part000.csv\n",
      "DEBUG:biocypher:Translating edges to BioCypher\n",
      "DEBUG:biocypher:Finished translating edges to BioCypher.\n",
      "DEBUG:biocypher:Writing edge CSV from generator.\n",
      "INFO -- Writing 650 entries to VariantToGeneAssociation-part000.csv\n",
      "INFO:biocypher:Writing 650 entries to VariantToGeneAssociation-part000.csv\n",
      "INFO -- Writing 650 entries to PatientHasVariant-part000.csv\n",
      "INFO:biocypher:Writing 650 entries to PatientHasVariant-part000.csv\n",
      "INFO -- Writing test import call to `/Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/neo4j-admin-import-call.sh`.\n",
      "INFO:biocypher:Writing test import call to `/Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/neo4j-admin-import-call.sh`.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/neo4j-admin-import-call.sh\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/equaglio/Decider/Ontoweaver/2WeeksProject/2WP_1db/biocypher-out/20250417175721/neo4j-admin-import-call.sh\n"
     ]
    }
   ],
   "source": [
    "#import decider_genetics.adapters.preprocessing_dataframes as preprocess_df\n",
    "#import decider_genetics.adapters.preprocessing_variants as preprocess_snv\n",
    "\n",
    "\n",
    "bc = BioCypher(\n",
    "    biocypher_config_path=\"config/biocypher_config.yaml\",\n",
    "    schema_config_path=\"config/schema_config_1.yaml\"\n",
    ")\n",
    "\n",
    "# Actually extract data.\n",
    "nodes = []\n",
    "edges = []\n",
    "\n",
    "data_mappings = {}\n",
    "\n",
    "logging.info(f\"Weave synthetic variants data\")\n",
    "\n",
    "#variants_df = pd.read_csv('data/', sep='\\t')\n",
    "\n",
    "mapping_file = \"./decider_genetics/adapters/my_mapping_1_variants.yaml\"\n",
    "with open(mapping_file) as fd:\n",
    "    conf = yaml.full_load(fd)\n",
    "\n",
    "adapter = ontoweaver.tabular.extract_table(df=variants_df, config=conf,separator = None, affix= \"none\")\n",
    "\n",
    "nodes += adapter.nodes\n",
    "edges += adapter.edges\n",
    "\n",
    "logging.info(f\"Wove variants: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "\n",
    "# # Write everything.\n",
    "# n, e = ontoweaver.extract(data_mappings)\n",
    "# nodes += n\n",
    "# edges += e\n",
    "\n",
    "import_file = ontoweaver.reconciliate_write(nodes, edges, \"config/biocypher_config.yaml\", \"config/schema_config_1.yaml\", separator=\", \")\n",
    "# bc.write_schema_info(as_node=True)\n",
    "\n",
    "print(import_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "422c4090",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/equaglio/Decider/decider-genetics\n"
     ]
    }
   ],
   "source": [
    "! pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "612df44d-a55e-4ca0-b293-7055abac5bcf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34m20250426162526\u001b[m\u001b[m \u001b[34m20250429170857\u001b[m\u001b[m \u001b[34m20250429174328\u001b[m\u001b[m \u001b[34m20250429183224\u001b[m\u001b[m \u001b[34m20250429195129\u001b[m\u001b[m\n",
      "\u001b[34m20250426173219\u001b[m\u001b[m \u001b[34m20250429172231\u001b[m\u001b[m \u001b[34m20250429174351\u001b[m\u001b[m \u001b[34m20250429183405\u001b[m\u001b[m\n",
      "\u001b[34m20250429165625\u001b[m\u001b[m \u001b[34m20250429174307\u001b[m\u001b[m \u001b[34m20250429174419\u001b[m\u001b[m \u001b[34m20250429183537\u001b[m\u001b[m\n",
      "\n",
      "biocypher-out/20250426162526:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv           neo4j-admin-import-call.sh-e\n",
      "\n",
      "biocypher-out/20250426173219:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv           neo4j-admin-import-call.sh-e\n",
      "\n",
      "biocypher-out/20250429165625:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv           neo4j-admin-import-call.sh-e\n",
      "\n",
      "biocypher-out/20250429170857:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv\n",
      "\n",
      "biocypher-out/20250429172231:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv\n",
      "\n",
      "biocypher-out/20250429174307:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv\n",
      "\n",
      "biocypher-out/20250429174328:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv\n",
      "\n",
      "biocypher-out/20250429174351:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv\n",
      "\n",
      "biocypher-out/20250429174419:\n",
      "Patient-header.csv                   VariantToGeneAssociation-header.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-header.csv         neo4j-admin-import-call.sh\n",
      "PatientHasVariant-part000.csv\n",
      "\n",
      "biocypher-out/20250429183224:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv\n",
      "\n",
      "biocypher-out/20250429183405:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv\n",
      "\n",
      "biocypher-out/20250429183537:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv\n",
      "\n",
      "biocypher-out/20250429195129:\n",
      "Patient-header.csv                   SequenceVariant-part000.csv\n",
      "Patient-part000.csv                  VariantToGeneAssociation-header.csv\n",
      "PatientHasVariant-header.csv         VariantToGeneAssociation-part000.csv\n",
      "PatientHasVariant-part000.csv        neo4j-admin-import-call.sh\n",
      "SequenceVariant-header.csv\n"
     ]
    }
   ],
   "source": [
    "! ls -R biocypher-out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "43f685de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Neo4j is not running.\n"
     ]
    }
   ],
   "source": [
    "! neo4j-admin server stop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5053d458-218d-4afe-ac07-907090e367fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Neo4j is not running.\n"
     ]
    }
   ],
   "source": [
    "! neo4j stop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5c76509a-a131-4160-9b4c-a0092d476066",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DO: substitute timestamp folder with the (last) one in `biocypher-out`\n",
    "! sed -i -e 's,bin/neo4j-admin,neo4j-admin,g' biocypher-out/20250429195129/neo4j-admin-import-call.sh  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "99bb31ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting to import, output will be saved to: /Users/equaglio/neo4j-community-2025.03.0/logs/neo4j-admin-import-2025-04-29.19.52.34.log\n",
      "Neo4j version: 2025.03.0\n",
      "Importing the contents of these files into /Users/equaglio/neo4j-community-2025.03.0/data/databases/test:\n",
      "Nodes:\n",
      "  /Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/SequenceVariant-header.csv\n",
      "  /Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/SequenceVariant-part000.csv\n",
      "  /Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/Patient-header.csv\n",
      "  /Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/Patient-part000.csv\n",
      "\n",
      "Relationships:\n",
      "  null:\n",
      "  /Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/VariantToGeneAssociation-header.csv\n",
      "  /Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/VariantToGeneAssociation-part000.csv\n",
      "  /Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/PatientHasVariant-header.csv\n",
      "  /Users/equaglio/Decider/decider-genetics/biocypher-out/20250429195129/PatientHasVariant-part000.csv\n",
      "\n",
      "\n",
      "Available resources:\n",
      "  Total machine memory: 24.00GiB\n",
      "  Free machine memory: 68.94MiB\n",
      "  Max heap memory : 5.333GiB\n",
      "  Max worker threads: 8\n",
      "  Configured max memory: 28.64MiB\n",
      "  High parallel IO: true\n",
      "\n",
      "WARNING: heap size 5.333GiB is unnecessarily large for completing this import.\n",
      "The abundant heap memory will leave less memory for off-heap importer caches. Suggested heap size is 1.000GiB\n",
      "Import starting 2025-04-29 19:52:36.003+0200\n",
      "  Estimated number of nodes: 670.00 \n",
      "  Estimated number of node properties: 1.34 k\n",
      "  Estimated number of relationships: 1.30 k\n",
      "  Estimated number of relationship properties: 0.00 \n",
      "  Estimated disk space usage: 82.46KiB\n",
      "  Estimated required memory usage: 71.51MiB\n",
      "\n",
      "(1/4) Node import 2025-04-29 19:52:36.032+0200\n",
      "  Estimated number of nodes: 670.00 \n",
      "  Estimated disk space usage: 39.30KiB\n",
      "  Estimated required memory usage: 71.51MiB\n",
      "-......... .......... .......... .......... ..........   5% ∆136ms [136ms]\n",
      ".......... .......... .......... .......... ..........  10% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  15% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  20% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  25% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  30% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  35% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  40% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  45% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  50% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  55% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  60% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  65% ∆0ms [137ms]\n",
      ".......... .......... .......... .......... ..........  70% ∆0ms [138ms]\n",
      ".......... .......... .......... .......... ..........  75% ∆0ms [138ms]\n",
      ".......... .......... .......... .......... ..........  80% ∆0ms [138ms]\n",
      ".......... .......... .......... .......... ..........  85% ∆0ms [138ms]\n",
      ".......... .......... .......... .......... ..........  90% ∆0ms [138ms]\n",
      ".......... .......... .......... .......... ..........  95% ∆0ms [138ms]\n",
      ".......... .......... .......... .......... .......... 100% ∆0ms [138ms]\n",
      "Node import COMPLETED in 147ms\n",
      "\n",
      "(2/4) Relationship import 2025-04-29 19:52:36.179+0200\n",
      "  Estimated number of relationships: 1.30 k\n",
      "  Estimated disk space usage: 43.16KiB\n",
      "  Estimated required memory usage: 75.50MiB\n",
      ".......... .......... .......... .......... ..........   5% ∆26ms [26ms]\n",
      ".......... .......... .......... .......... ..........  10% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  15% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  20% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  25% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  30% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  35% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  40% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  45% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  50% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  55% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  60% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  65% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  70% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  75% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  80% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  85% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  90% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... ..........  95% ∆0ms [26ms]\n",
      ".......... .......... .......... .......... .......... 100% ∆0ms [26ms]\n",
      "Relationship import COMPLETED in 27ms\n",
      "\n",
      "(3/4) Relationship linking 2025-04-29 19:52:36.206+0200\n",
      "  Estimated required memory usage: 71.51MiB\n",
      "-......... .......... .......... .......... ..........   5% ∆80ms [80ms]\n",
      ".......... .......... .......... .......... ..........  10% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  15% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  20% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  25% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  30% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  35% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  40% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  45% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  50% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  55% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  60% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  65% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  70% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  75% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  80% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  85% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  90% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... ..........  95% ∆0ms [80ms]\n",
      ".......... .......... .......... .......... .......... 100% ∆0ms [80ms]\n",
      "Relationship linking COMPLETED in 80ms\n",
      "\n",
      "(4/4) Post processing 2025-04-29 19:52:36.286+0200\n",
      "  Estimated required memory usage: 71.50MiB\n",
      "-......... .......... .......... .......... ..........   5% ∆242ms [242ms]\n",
      ".......... .......... .......... .......... ..........  10% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  15% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  20% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  25% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  30% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  35% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  40% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  45% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  50% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  55% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  60% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  65% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  70% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  75% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  80% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  85% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  90% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... ..........  95% ∆0ms [242ms]\n",
      ".......... .......... .......... .......... .......... 100% ∆0ms [242ms]\n",
      "Post processing COMPLETED in 243ms\n",
      "\n",
      "\n",
      "IMPORT DONE in 683ms. \n",
      "Imported:\n",
      "  670 nodes\n",
      "  650 relationships\n",
      "  1340 properties\n",
      "Peak memory usage: 91.02MiB\n"
     ]
    }
   ],
   "source": [
    "! sh biocypher-out/20250429195129/neo4j-admin-import-call.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ea4b91d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directories in use:\n",
      "home:         /Users/equaglio/neo4j-community-2025.03.0\n",
      "config:       /Users/equaglio/neo4j-community-2025.03.0/conf\n",
      "logs:         /Users/equaglio/neo4j-community-2025.03.0/logs\n",
      "plugins:      /Users/equaglio/neo4j-community-2025.03.0/plugins\n",
      "import:       /Users/equaglio/neo4j-community-2025.03.0/import\n",
      "data:         /Users/equaglio/neo4j-community-2025.03.0/data\n",
      "certificates: /Users/equaglio/neo4j-community-2025.03.0/certificates\n",
      "licenses:     /Users/equaglio/neo4j-community-2025.03.0/licenses\n",
      "run:          /Users/equaglio/neo4j-community-2025.03.0/run\n",
      "Starting Neo4j.\n",
      "2025-04-29 14:57:47.946+0000 INFO  Logging config in use: File '/Users/equaglio/neo4j-community-2025.03.0/conf/user-logs.xml'\n",
      "2025-04-29 14:57:47.951+0000 INFO  Starting...\n",
      "2025-04-29 14:57:48.537+0000 INFO  This instance is ServerId{865d2652} (865d2652-49d9-4958-b8e4-33516e9c653f)\n",
      "2025-04-29 14:57:49.125+0000 INFO  ======== Neo4j 2025.03.0 ========\n",
      "2025-04-29 14:57:50.065+0000 INFO  Anonymous Usage Data is being sent to Neo4j, see https://neo4j.com/docs/usage-data/\n",
      "2025-04-29 14:57:50.226+0000 INFO  Bolt enabled on localhost:7687.\n",
      "2025-04-29 14:57:50.547+0000 INFO  HTTP enabled on localhost:7474.\n",
      "2025-04-29 14:57:50.547+0000 INFO  Remote interface available at http://localhost:7474/\n",
      "2025-04-29 14:57:50.549+0000 INFO  id: BD004600D2AA4ABAFF38612F51B57D210B7156B06181FB993E1BCCE23A627188\n",
      "2025-04-29 14:57:50.549+0000 INFO  name: system\n",
      "2025-04-29 14:57:50.549+0000 INFO  creationDate: 2025-04-14T16:14:55.359Z\n",
      "2025-04-29 14:57:50.549+0000 INFO  Started.\n",
      "^C\n",
      "2025-04-29 15:01:08.002+0000 INFO  Neo4j Server shutdown initiated by request\n",
      "2025-04-29 15:01:08.004+0000 INFO  Stopping...\n"
     ]
    }
   ],
   "source": [
    "#! neo4j-admin server start # this for some reason does not work\n",
    "! neo4j console # this one does"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "696d8970-3e69-4007-988e-927ebb7e1825",
   "metadata": {},
   "source": [
    "`MATCH (n)-[r]->(m) RETURN n, r, m` on neo4j to visualize the SKG"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "796161a9-a793-4bd8-989e-f2a8e70aef59",
   "metadata": {},
   "source": [
    "Interrupt execution of previous block to gently shutdown neo4j server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "32f6915e-61fb-439d-a5bf-a87d5c2e1d75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Neo4j is not running.\n",
      "Run with '--verbose' for a more detailed error message.\n"
     ]
    }
   ],
   "source": [
    "! neo4j status # to check, after some seconds, the effect of shutdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "eb2e20cd-7251-4028-bcb1-3af7c2c9db46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# in case shut down didnt work, run \n",
    "# ! ps aux | grep neo4j\n",
    "# and then \n",
    "# kill <PID>\n",
    "# where <PID> is the process identifier of the neo4j session"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a22a09f7-baa3-4d23-8f6d-5a748b3882d0",
   "metadata": {},
   "source": [
    "## Now let's add properties to those nodes\n",
    "\n",
    "Patients properties are stored in `data/synthetic_clinical.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "f4834d0c-ecef-494e-94e1-89be4cd9468e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Age</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Chemotherapy cycles</th>\n",
       "      <th>Maintenance</th>\n",
       "      <th>PARPi</th>\n",
       "      <th>Primary outcome</th>\n",
       "      <th>BRCA mutation</th>\n",
       "      <th>HR deficient</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>patient1</td>\n",
       "      <td>72</td>\n",
       "      <td>21,484375</td>\n",
       "      <td>IVB</td>\n",
       "      <td>PDS</td>\n",
       "      <td>6.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Partial Response</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HRD negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>patient2</td>\n",
       "      <td>64</td>\n",
       "      <td>28,62879239</td>\n",
       "      <td>IIB</td>\n",
       "      <td>PDS</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Partial Response</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>patient3</td>\n",
       "      <td>70</td>\n",
       "      <td>26,83864765</td>\n",
       "      <td>IVA</td>\n",
       "      <td>NACT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Partial Response</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>patient4</td>\n",
       "      <td>78</td>\n",
       "      <td>23,05456246</td>\n",
       "      <td>IVA</td>\n",
       "      <td>NACT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Letrozole</td>\n",
       "      <td>No</td>\n",
       "      <td>Partial Response</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HRD positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>patient5</td>\n",
       "      <td>75</td>\n",
       "      <td>20,703125</td>\n",
       "      <td>IVB</td>\n",
       "      <td>PDS</td>\n",
       "      <td>6.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Progressive Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HRD negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Patient  Age          BMI Stage Treatment  Chemotherapy cycles Maintenance PARPi      Primary outcome BRCA mutation  HR deficient\n",
       "0  patient1   72    21,484375   IVB       PDS                  6.0          No    No     Partial Response           NaN  HRD negative\n",
       "1  patient2   64  28,62879239   IIB       PDS                  6.0         NaN    No     Partial Response           Yes           NaN\n",
       "2  patient3   70  26,83864765   IVA      NACT                  6.0          No   Yes     Partial Response           NaN           NaN\n",
       "3  patient4   78  23,05456246   IVA      NACT                  6.0   Letrozole    No     Partial Response           NaN  HRD positive\n",
       "4  patient5   75    20,703125   IVB       PDS                  6.0          No    No  Progressive Disease           NaN  HRD negative"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df = pd.read_table('data/synthetic_clinical.csv', sep=';')\n",
    "clinical_df.head(n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fda2b840-6688-4c35-8e31-561bf23b93c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edit the biocypher config to point to the new schema_config file\n",
    "\n",
    "with open('config/biocypher_config.yaml', 'r') as f:\n",
    "    bc = yaml.safe_load(f)\n",
    "\n",
    "bc['biocypher']['schema_config_path'] = 'schema_config_2.yaml'\n",
    "\n",
    "with open('config/biocypher_config.yaml', 'w') as f:\n",
    "    yaml.dump(bc, f, default_flow_style=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9135b53a-1968-4c4b-bc8c-90db684aba1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO -- Running BioCypher with schema configuration from config/schema_config_2.yaml.\n",
      "INFO:biocypher:Running BioCypher with schema configuration from config/schema_config_2.yaml.\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "'NoneType' object is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[42]\u001b[39m\u001b[32m, line 39\u001b[39m\n\u001b[32m     36\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(mapping_file) \u001b[38;5;28;01mas\u001b[39;00m fd:\n\u001b[32m     37\u001b[39m     conf = yaml.full_load(fd)\n\u001b[32m---> \u001b[39m\u001b[32m39\u001b[39m adapter = \u001b[43montoweaver\u001b[49m\u001b[43m.\u001b[49m\u001b[43mtabular\u001b[49m\u001b[43m.\u001b[49m\u001b[43mextract_table\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf\u001b[49m\u001b[43m=\u001b[49m\u001b[43mclinical_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m=\u001b[49m\u001b[43mconf\u001b[49m\u001b[43m,\u001b[49m\u001b[43mseparator\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maffix\u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mnone\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m     41\u001b[39m nodes += adapter.nodes\n\u001b[32m     42\u001b[39m edges += adapter.edges\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Decider/Ontoweaver/src/ontoweaver/tabular.py:494\u001b[39m, in \u001b[36mextract_table\u001b[39m\u001b[34m(df, config, parallel_mapping, module, affix, separator, raise_errors)\u001b[39m\n\u001b[32m    477\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    478\u001b[39m \u001b[33;03mProxy function for extracting from a table all nodes, edges and properties\u001b[39;00m\n\u001b[32m    479\u001b[39m \u001b[33;03mthat are defined in a PandasAdapter configuration.\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m    491\u001b[39m \u001b[33;03m    PandasAdapter: The configured adapter.\u001b[39;00m\n\u001b[32m    492\u001b[39m \u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    493\u001b[39m parser = YamlParser(config, module, raise_errors = raise_errors)\n\u001b[32m--> \u001b[39m\u001b[32m494\u001b[39m mapping = \u001b[43mparser\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    496\u001b[39m adapter = PandasAdapter(\n\u001b[32m    497\u001b[39m     df,\n\u001b[32m    498\u001b[39m     *mapping,\n\u001b[32m   (...)\u001b[39m\u001b[32m    502\u001b[39m     raise_errors = raise_errors,\n\u001b[32m    503\u001b[39m )\n\u001b[32m    505\u001b[39m adapter.run()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Decider/Ontoweaver/src/ontoweaver/tabular.py:875\u001b[39m, in \u001b[36m__call__\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    872\u001b[39m     \u001b[38;5;28;01massert\u001b[39;00m (\u001b[38;5;28mtype\u001b[39m(object_types) == \u001b[38;5;28mstr\u001b[39m)\n\u001b[32m    873\u001b[39m     object_types = [object_types]\n\u001b[32m--> \u001b[39m\u001b[32m875\u001b[39m column_names = \u001b[38;5;28mself\u001b[39m.get(\u001b[38;5;28mself\u001b[39m.k_columns, pconfig=field_dict)\n\u001b[32m    876\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m column_names != \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(column_names) != \u001b[38;5;28mlist\u001b[39m:\n\u001b[32m    877\u001b[39m     logger.debug(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[33mDeclared singular column `\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcolumn_names\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m`\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[31mTypeError\u001b[39m: 'NoneType' object is not iterable"
     ]
    }
   ],
   "source": [
    "#import decider_genetics.adapters.preprocessing_dataframes as preprocess_df\n",
    "#import decider_genetics.adapters.preprocessing_variants as preprocess_snv\n",
    "\n",
    "\n",
    "bc = BioCypher(\n",
    "    biocypher_config_path=\"config/biocypher_config.yaml\",\n",
    "    schema_config_path=\"config/schema_config_2.yaml\"\n",
    ")\n",
    "\n",
    "# Actually extract data.\n",
    "nodes = []\n",
    "edges = []\n",
    "\n",
    "data_mappings = {}\n",
    "\n",
    "logging.info(f\"Weave synthetic variants data\")\n",
    "\n",
    "#variants_df = pd.read_csv('data/', sep='\\t')\n",
    "\n",
    "mapping_file = \"./decider_genetics/adapters/my_mapping_2_variants.yaml\"\n",
    "with open(mapping_file) as fd:\n",
    "    conf = yaml.full_load(fd)\n",
    "\n",
    "adapter = ontoweaver.tabular.extract_table(df=variants_df, config=conf,separator = \"\\t\", affix= \"none\")\n",
    "\n",
    "nodes += adapter.nodes\n",
    "edges += adapter.edges\n",
    "\n",
    "logging.info(f\"Wove variants: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "\n",
    "\n",
    "\n",
    "logging.info(f\"Weave synthetic clinical data...\")\n",
    "\n",
    "mapping_file = \"./decider_genetics/adapters/my_mapping_2_clinical.yaml\"\n",
    "with open(mapping_file) as fd:\n",
    "    conf = yaml.full_load(fd)\n",
    "\n",
    "adapter = ontoweaver.tabular.extract_table(df=clinical_df, config=conf,separator = \"\\t\", affix= \"none\")\n",
    "\n",
    "nodes += adapter.nodes\n",
    "edges += adapter.edges\n",
    "\n",
    "logging.info(f\"Wove Clinical: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "\n",
    "\n",
    "# # Write everything.\n",
    "# n, e = ontoweaver.extract(data_mappings)\n",
    "# nodes += n\n",
    "# edges += e\n",
    "\n",
    "import_file = ontoweaver.reconciliate_write(nodes, edges, \"config/biocypher_config.yaml\", \"config/schema_config_2.yaml\", separator=\", \")\n",
    "# bc.write_schema_info(as_node=True)\n",
    "\n",
    "print(import_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53f46bdc-4485-4013-b48e-3636ae8a29fe",
   "metadata": {},
   "source": [
    "TO FIX!!! The previous attempt gives error, either with or without properties (tried cos there was apparently a bug in property parsing but later was fixed and i pulled onto my local ontoweaver repo so it shoud not be the problem...)\n",
    "\n",
    "Same error also in the third schema version. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5609831d-aafd-451b-9dc9-8bbf56a3e3f6",
   "metadata": {},
   "source": [
    "## And finally let's add a CopyNumberAlteration node connecting Patient and Gene as well.\n",
    "Those alterations are store in Patients `data/synthetic_cns.csv`. Such tables store also information about the chromosome of location of genes present in the `synthetic_variants.csv` table, so we add it as gene property."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "bf54927a-1c7d-4996-bafb-fb57caa9df5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>band</th>\n",
       "      <th>type</th>\n",
       "      <th>sample</th>\n",
       "      <th>nProbesCr</th>\n",
       "      <th>nProbesAf</th>\n",
       "      <th>logR</th>\n",
       "      <th>baf</th>\n",
       "      <th>nAraw</th>\n",
       "      <th>nBraw</th>\n",
       "      <th>nMajor</th>\n",
       "      <th>nMinor</th>\n",
       "      <th>purifiedLogR</th>\n",
       "      <th>purifiedBaf</th>\n",
       "      <th>purifiedLoh</th>\n",
       "      <th>CNstatus</th>\n",
       "      <th>LOHstatus</th>\n",
       "      <th>minPurifiedLogR</th>\n",
       "      <th>maxPurifiedLogR</th>\n",
       "      <th>breaksInGene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000168685</td>\n",
       "      <td>IL7R</td>\n",
       "      <td>chr5</td>\n",
       "      <td>35852695</td>\n",
       "      <td>35879603</td>\n",
       "      <td>1</td>\n",
       "      <td>p13.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>patient1</td>\n",
       "      <td>7595</td>\n",
       "      <td>5440</td>\n",
       "      <td>0.9068</td>\n",
       "      <td>0.3309</td>\n",
       "      <td>9.5999</td>\n",
       "      <td>4.1648</td>\n",
       "      <td>10.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.1350</td>\n",
       "      <td>0.3026</td>\n",
       "      <td>0.3949</td>\n",
       "      <td>AMP</td>\n",
       "      <td>HET</td>\n",
       "      <td>1.1350</td>\n",
       "      <td>1.1350</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000119772</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25227855</td>\n",
       "      <td>25342590</td>\n",
       "      <td>-1</td>\n",
       "      <td>p23.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>patient1</td>\n",
       "      <td>80852</td>\n",
       "      <td>45350</td>\n",
       "      <td>0.0811</td>\n",
       "      <td>0.4819</td>\n",
       "      <td>2.7730</td>\n",
       "      <td>2.0427</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.3761</td>\n",
       "      <td>0.4242</td>\n",
       "      <td>0.1517</td>\n",
       "      <td>Normal</td>\n",
       "      <td>HET</td>\n",
       "      <td>0.3761</td>\n",
       "      <td>0.3761</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000068078</td>\n",
       "      <td>FGFR3</td>\n",
       "      <td>chr4</td>\n",
       "      <td>1793293</td>\n",
       "      <td>1808872</td>\n",
       "      <td>1</td>\n",
       "      <td>p16.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>patient1</td>\n",
       "      <td>11371</td>\n",
       "      <td>8948</td>\n",
       "      <td>0.6256</td>\n",
       "      <td>0.4086</td>\n",
       "      <td>3.2568</td>\n",
       "      <td>2.1358</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.7372</td>\n",
       "      <td>0.3961</td>\n",
       "      <td>0.2079</td>\n",
       "      <td>Normal</td>\n",
       "      <td>HET</td>\n",
       "      <td>0.7372</td>\n",
       "      <td>0.7372</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000185920</td>\n",
       "      <td>PTCH1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>95442980</td>\n",
       "      <td>95517057</td>\n",
       "      <td>-1</td>\n",
       "      <td>q22.32</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>patient1</td>\n",
       "      <td>31710</td>\n",
       "      <td>11538</td>\n",
       "      <td>-0.1153</td>\n",
       "      <td>0.4152</td>\n",
       "      <td>3.1927</td>\n",
       "      <td>2.2095</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-0.1234</td>\n",
       "      <td>0.4090</td>\n",
       "      <td>0.1820</td>\n",
       "      <td>Normal</td>\n",
       "      <td>HET</td>\n",
       "      <td>-0.1234</td>\n",
       "      <td>-0.1234</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000106462</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>148807383</td>\n",
       "      <td>148884321</td>\n",
       "      <td>-1</td>\n",
       "      <td>q36.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>patient1</td>\n",
       "      <td>7948</td>\n",
       "      <td>5736</td>\n",
       "      <td>0.3978</td>\n",
       "      <td>0.4968</td>\n",
       "      <td>2.5976</td>\n",
       "      <td>2.5558</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.5297</td>\n",
       "      <td>0.4959</td>\n",
       "      <td>0.0081</td>\n",
       "      <td>Normal</td>\n",
       "      <td>HET</td>\n",
       "      <td>0.5297</td>\n",
       "      <td>0.5297</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                ID    Gene   chr      start        end  strand    band            type    sample  nProbesCr  nProbesAf    logR     baf   nAraw  \\\n",
       "0  ENSG00000168685    IL7R  chr5   35852695   35879603       1   p13.2  protein_coding  patient1       7595       5440  0.9068  0.3309  9.5999   \n",
       "1  ENSG00000119772  DNMT3A  chr2   25227855   25342590      -1   p23.3  protein_coding  patient1      80852      45350  0.0811  0.4819  2.7730   \n",
       "2  ENSG00000068078   FGFR3  chr4    1793293    1808872       1   p16.3  protein_coding  patient1      11371       8948  0.6256  0.4086  3.2568   \n",
       "3  ENSG00000185920   PTCH1  chr9   95442980   95517057      -1  q22.32  protein_coding  patient1      31710      11538 -0.1153  0.4152  3.1927   \n",
       "4  ENSG00000106462    EZH2  chr7  148807383  148884321      -1   q36.1  protein_coding  patient1       7948       5736  0.3978  0.4968  2.5976   \n",
       "\n",
       "    nBraw  nMajor  nMinor  purifiedLogR  purifiedBaf  purifiedLoh CNstatus LOHstatus  minPurifiedLogR  maxPurifiedLogR  breaksInGene  \n",
       "0  4.1648    10.0     4.0        1.1350       0.3026       0.3949      AMP       HET           1.1350           1.1350             0  \n",
       "1  2.0427     3.0     2.0        0.3761       0.4242       0.1517   Normal       HET           0.3761           0.3761             0  \n",
       "2  2.1358     3.0     2.0        0.7372       0.3961       0.2079   Normal       HET           0.7372           0.7372             0  \n",
       "3  2.2095     3.0     2.0       -0.1234       0.4090       0.1820   Normal       HET          -0.1234          -0.1234             0  \n",
       "4  2.5558     3.0     3.0        0.5297       0.4959       0.0081   Normal       HET           0.5297           0.5297             0  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cna_df = pd.read_table('data/synthetic_cna.csv', sep='\\t')\n",
    "cna_df.head(n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3c5e5233-3462-4832-bd4e-54522c4c22e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edit the biocypher config to point to the new schema_config file\n",
    "with open('config/biocypher_config.yaml', 'r') as f:\n",
    "        bc = yaml.safe_load(f)\n",
    "        bc['biocypher']['schema_config_path'] = 'schema_config_3.yaml'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c4f6dcd3-d6b1-4e9e-9adb-90eaf18beb34",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO -- Running BioCypher with schema configuration from config/schema_config_3.yaml.\n",
      "INFO:biocypher:Running BioCypher with schema configuration from config/schema_config_3.yaml.\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'items'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[47]\u001b[39m\u001b[32m, line 54\u001b[39m\n\u001b[32m     51\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(mapping_file) \u001b[38;5;28;01mas\u001b[39;00m fd:\n\u001b[32m     52\u001b[39m     conf = yaml.full_load(fd)\n\u001b[32m---> \u001b[39m\u001b[32m54\u001b[39m adapter = \u001b[43montoweaver\u001b[49m\u001b[43m.\u001b[49m\u001b[43mtabular\u001b[49m\u001b[43m.\u001b[49m\u001b[43mextract_table\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcna_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m=\u001b[49m\u001b[43mconf\u001b[49m\u001b[43m,\u001b[49m\u001b[43mseparator\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maffix\u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mnone\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m     56\u001b[39m nodes += adapter.nodes\n\u001b[32m     57\u001b[39m edges += adapter.edges\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Decider/Ontoweaver/src/ontoweaver/tabular.py:494\u001b[39m, in \u001b[36mextract_table\u001b[39m\u001b[34m(df, config, parallel_mapping, module, affix, separator, raise_errors)\u001b[39m\n\u001b[32m    477\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    478\u001b[39m \u001b[33;03mProxy function for extracting from a table all nodes, edges and properties\u001b[39;00m\n\u001b[32m    479\u001b[39m \u001b[33;03mthat are defined in a PandasAdapter configuration.\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m    491\u001b[39m \u001b[33;03m    PandasAdapter: The configured adapter.\u001b[39;00m\n\u001b[32m    492\u001b[39m \u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    493\u001b[39m parser = YamlParser(config, module, raise_errors = raise_errors)\n\u001b[32m--> \u001b[39m\u001b[32m494\u001b[39m mapping = \u001b[43mparser\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    496\u001b[39m adapter = PandasAdapter(\n\u001b[32m    497\u001b[39m     df,\n\u001b[32m    498\u001b[39m     *mapping,\n\u001b[32m   (...)\u001b[39m\u001b[32m    502\u001b[39m     raise_errors = raise_errors,\n\u001b[32m    503\u001b[39m )\n\u001b[32m    505\u001b[39m adapter.run()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Decider/Ontoweaver/src/ontoweaver/tabular.py:876\u001b[39m, in \u001b[36mYamlParser.__call__\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    874\u001b[39m logger.debug(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mParse properties...\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    875\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m n_transformer,transformer_types \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(transformers_list):\n\u001b[32m--> \u001b[39m\u001b[32m876\u001b[39m     \u001b[38;5;28;01mfor\u001b[39;00m transformer_type, field_dict \u001b[38;5;129;01min\u001b[39;00m \u001b[43mtransformer_types\u001b[49m\u001b[43m.\u001b[49m\u001b[43mitems\u001b[49m():\n\u001b[32m    877\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m field_dict:\n\u001b[32m    878\u001b[39m             \u001b[38;5;28mself\u001b[39m.error(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mThere is no field for the \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mn_transformer\u001b[38;5;132;01m}\u001b[39;00m\u001b[33mth transformer: \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtransformer_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m,\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    879\u001b[39m                        \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33m did you forget an indentation?\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mtransformers\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m    880\u001b[39m                        n_transformer, exception = exceptions.MissingFieldError)\n",
      "\u001b[31mAttributeError\u001b[39m: 'str' object has no attribute 'items'"
     ]
    }
   ],
   "source": [
    "#import decider_genetics.adapters.preprocessing_dataframes as preprocess_df\n",
    "#import decider_genetics.adapters.preprocessing_variants as preprocess_snv\n",
    "\n",
    "\n",
    "bc = BioCypher(\n",
    "    biocypher_config_path=\"config/biocypher_config.yaml\",\n",
    "    schema_config_path=\"config/schema_config_3.yaml\"\n",
    ")\n",
    "\n",
    "# Actually extract data.\n",
    "nodes = []\n",
    "edges = []\n",
    "\n",
    "data_mappings = {}\n",
    "\n",
    "logging.info(f\"Weave synthetic variants data\")\n",
    "\n",
    "#variants_df = pd.read_csv('data/', sep='\\t')\n",
    "\n",
    "mapping_file = \"./decider_genetics/adapters/my_mapping_3_variants.yaml\"\n",
    "with open(mapping_file) as fd:\n",
    "    conf = yaml.full_load(fd)\n",
    "\n",
    "adapter = ontoweaver.tabular.extract_table(df=variants_df, config=conf,separator = None, affix= \"none\")\n",
    "\n",
    "nodes += adapter.nodes\n",
    "edges += adapter.edges\n",
    "\n",
    "logging.info(f\"Wove variants: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "\n",
    "\n",
    "'''\n",
    "logging.info(f\"Weave synthetic clinical data...\")\n",
    "\n",
    "mapping_file = \"./decider_genetics/adapters/my_mapping_3_clinical.yaml\"\n",
    "with open(mapping_file) as fd:\n",
    "    conf = yaml.full_load(fd)\n",
    "\n",
    "adapter = ontoweaver.tabular.extract_table(df=clinical_df, config=conf,separator = \";\", affix= \"none\")\n",
    "\n",
    "nodes += adapter.nodes\n",
    "edges += adapter.edges\n",
    "\n",
    "logging.info(f\"Wove Clinical: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "'''\n",
    "\n",
    "\n",
    "logging.info(f\"Weave synthetic copy number alteration data...\")\n",
    "\n",
    "mapping_file = \"./decider_genetics/adapters/my_mapping_3_cna.yaml\"\n",
    "with open(mapping_file) as fd:\n",
    "    conf = yaml.full_load(fd)\n",
    "\n",
    "adapter = ontoweaver.tabular.extract_table(df=cna_df, config=conf,separator = None, affix= \"none\")\n",
    "\n",
    "nodes += adapter.nodes\n",
    "edges += adapter.edges\n",
    "\n",
    "logging.info(f\"Wove CNAs: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "\n",
    "# # Write everything.\n",
    "# n, e = ontoweaver.extract(data_mappings)\n",
    "# nodes += n\n",
    "# edges += e\n",
    "\n",
    "import_file = ontoweaver.reconciliate_write(nodes, edges, \"config/biocypher_config.yaml\", \"config/schema_config_3.yaml\", separator=\", \")\n",
    "# bc.write_schema_info(as_node=True)\n",
    "\n",
    "print(import_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0000c35-13aa-4644-b6c6-3964d61d800e",
   "metadata": {},
   "source": [
    "#!/usr/bin/env python3\n",
    "import argparse\n",
    "import logging\n",
    "import pandas as pd\n",
    "import yaml\n",
    "\n",
    "from biocypher import BioCypher\n",
    "import decider_genetics.adapters.preprocessing_dataframes as preprocess_df\n",
    "import decider_genetics.adapters.preprocessing_variants as preprocess_snv\n",
    "import ontoweaver\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    usage = f\"Extract nodes and edges from CSV tables from synthetic clinical, copy number variations and single nucleotide variants and prepare a knowledge graph import script.\"\n",
    "    parser = argparse.ArgumentParser(\n",
    "        description=usage)\n",
    "\n",
    "    parser.add_argument(\"-cli\", \"--synthetic_clinical\", metavar=\"CSV\", nargs=\"+\",\n",
    "                        help=\"Extract from the synthetic clinical CSV file.\")\n",
    "\n",
    "    parser.add_argument(\"-cna\", \"--synthetic_cna\", metavar=\"CSV\", nargs=\"+\",\n",
    "                        help=\"Extract from the synthetic copy number alterations CSV file.\")\n",
    "\n",
    "    parser.add_argument(\"-snv\", \"--synthetic_variants\", metavar=\"CSV\", nargs=\"+\",\n",
    "                        help=\"Extract from the synthetic single nucleotide variants CSV file.\")\n",
    "    \n",
    "    parser.add_argument(\"-o\", \"--oncokb\", metavar=\"CSV\", nargs=\"+\",\n",
    "                        help=\"Extract from an OncoKB CSV file.\")\n",
    "    \n",
    "    levels = {\n",
    "        \"DEBUG\": logging.DEBUG,\n",
    "        \"INFO\": logging.INFO,\n",
    "        \"WARNING\": logging.WARNING,\n",
    "        \"ERROR\": logging.ERROR,\n",
    "        \"CRITICAL\": logging.CRITICAL\n",
    "    }\n",
    "\n",
    "    parser.add_argument(\"-v\", \"--verbose\", choices = levels.keys(), default = \"WARNING\",\n",
    "                        help=\"Set the verbose level (default: %(default)s).\")\n",
    "    \n",
    "    asked = parser.parse_args()\n",
    "    bc = BioCypher(\n",
    "        biocypher_config_path=\"config/biocypher_config.yaml\",\n",
    "        schema_config_path=\"config/schema_config.yaml\"\n",
    "    )\n",
    "\n",
    "    # Actually extract data.\n",
    "    nodes = []\n",
    "    edges = []\n",
    "\n",
    "    data_mappings = {}\n",
    "\n",
    "    # Extract from databases not requiring preprocessing.\n",
    "    if asked.synthetic_clinical:\n",
    "        logging.info(f\"Weave synthetic clinical data...\")\n",
    "        \n",
    "        clinical_df = pd.read_csv(asked.synthetic_clinical[0], sep=';')\n",
    "        preprocessed_clinical_df = preprocess_df.preprocess_clinical(clinical_df)\n",
    "\n",
    "        mapping_file = \"./decider_genetics/adapters/clinical.yaml\"\n",
    "        with open(mapping_file) as fd:\n",
    "            conf = yaml.full_load(fd)\n",
    "\n",
    "        adapter = ontoweaver.tabular.extract_all(df=preprocessed_clinical_df, config=conf,separator = None, affix= \"none\")\n",
    "\n",
    "        nodes += adapter.nodes\n",
    "        edges += adapter.edges\n",
    "\n",
    "        logging.info(f\"Wove Clinical: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "\n",
    "    if asked.synthetic_cna:\n",
    "        logging.info(f\"Weave synthetic copy number alterations data\")\n",
    "        \n",
    "        cna_df = pd.read_csv(asked.synthetic_cna[0], sep='\\t')\n",
    "        preprocessed_cna_df = preprocess_df.preprocess_cna(cna_df)\n",
    "\n",
    "        mapping_file = \"./decider_genetics/adapters/cna_genes.yaml\"\n",
    "        with open(mapping_file) as fd:\n",
    "            conf = yaml.full_load(fd)\n",
    "\n",
    "        adapter = ontoweaver.tabular.extract_all(df=preprocessed_cna_df, config=conf,separator = None, affix= \"none\")\n",
    "\n",
    "        nodes += adapter.nodes\n",
    "        edges += adapter.edges\n",
    "\n",
    "        logging.info(f\"Wove CNAs: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "\n",
    "    if asked.synthetic_variants:\n",
    "        logging.info(f\"Weave synthetic variants data\")\n",
    "\n",
    "        variants_df = pd.read_csv(asked.synthetic_variants[0], sep='\\t')\n",
    "        preprocessed_variants_df = preprocess_snv.preprocess_variants(variants_df)\n",
    "\n",
    "        mapping_file = \"./decider_genetics/adapters/variants.yaml\"\n",
    "        with open(mapping_file) as fd:\n",
    "            conf = yaml.full_load(fd)\n",
    "\n",
    "        adapter = ontoweaver.tabular.extract_all(df=preprocessed_variants_df, config=conf,separator = None, affix= \"none\")\n",
    "\n",
    "        nodes += adapter.nodes\n",
    "        edges += adapter.edges\n",
    "\n",
    "        logging.info(f\"Wove variants: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "    \n",
    "    if asked.oncokb:\n",
    "        logging.info(f\"Weave OncoKB data...\")\n",
    "        oncokb_df = pd.read_csv(asked.oncokb[0], sep='\\t')\n",
    "        preprocessed_oncokb_df = preprocess_df.preprocess_oncokb(oncokb_df)\n",
    "\n",
    "        mapping_file = \"./decider_genetics/adapters/oncokb.yaml\"\n",
    "        with open(mapping_file) as fd:\n",
    "            conf = yaml.full_load(fd)\n",
    "\n",
    "        adapter = ontoweaver.tabular.extract_all(df=preprocessed_oncokb_df, config=conf,separator = None, affix= \"none\")\n",
    "\n",
    "        nodes += adapter.nodes\n",
    "        edges += adapter.edges\n",
    "\n",
    "        logging.info(f\"Wove OncoKB: {len(nodes)} nodes, {len(edges)} edges.\")\n",
    "\n",
    "    # # Write everything.\n",
    "    # n, e = ontoweaver.extract(data_mappings)\n",
    "    # nodes += n\n",
    "    # edges += e\n",
    "\n",
    "    import_file = ontoweaver.reconciliate_write(nodes, edges, \"config/biocypher_config.yaml\", \"config/schema_config.yaml\", separator=\", \")\n",
    "    # bc.write_schema_info(as_node=True)\n",
    "\n",
    "    print(import_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "decider genetics poetry env",
   "language": "python",
   "name": "my-poetry-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
